internal/kafka/internal/handlers/kafka_promote.go

```diff
@@ -0,0 +1,76 @@
+package handlers
+
+import (
+	"net/http"
+
+	"github.com/bf2fc6cc711aee1a0c2a/kas-fleet-manager/internal/kafka/internal/api/dbapi"
+	"github.com/bf2fc6cc711aee1a0c2a/kas-fleet-manager/internal/kafka/internal/api/public"
+	"github.com/bf2fc6cc711aee1a0c2a/kas-fleet-manager/internal/kafka/internal/config"
+	"github.com/bf2fc6cc711aee1a0c2a/kas-fleet-manager/internal/kafka/internal/services"
+	"github.com/bf2fc6cc711aee1a0c2a/kas-fleet-manager/pkg/errors"
+	"github.com/bf2fc6cc711aee1a0c2a/kas-fleet-manager/pkg/handlers"
+	"github.com/bf2fc6cc711aee1a0c2a/kas-fleet-manager/pkg/services/authorization"
+	"github.com/gorilla/mux"
+)
+
+type kafkaPromoteHandler struct {
+	service                        services.KafkaService
+	authService                    authorization.Authorization // TODO keep or remove this depending on the other TODOs of the kafka promote handlers
+	kafkaConfig                    *config.KafkaConfig
+	kafkaPromotionValidatorFactory KafkaPromotionValidatorFactory
+}
+
+func NewKafkaPromoteHandler(service services.KafkaService, authService authorization.Authorization, kafkaConfig *config.KafkaConfig, kafkaPromotionValidatorFactory KafkaPromotionValidatorFactory) *kafkaPromoteHandler {
+	return &kafkaPromoteHandler{
+		service:                        service,
+		authService:                    authService,
+		kafkaConfig:                    kafkaConfig,
+		kafkaPromotionValidatorFactory: kafkaPromotionValidatorFactory,
+	}
+}
+
+const kafkaPromoteRouteVariableID = "id"
+
+func (h kafkaPromoteHandler) Promote(w http.ResponseWriter, r *http.Request) {
+	var kafkaPromotionRequest public.KafkaPromotionRequest
+	id := mux.Vars(r)[kafkaPromoteRouteVariableID]
+	ctx := r.Context()
+	kafkaRequest, kafkaGetError := h.service.Get(ctx, id)
+	validateKafkaFound := func() handlers.Validate {
+		return func() *errors.ServiceError {
+			return kafkaGetError
+		}
+	}
+	cfg := &handlers.HandlerConfig{
+		MarshalInto: &kafkaPromotionRequest,
+		Validate: []handlers.Validate{
+			handlers.ValidateMinLength(&kafkaPromotionRequest.DesiredKafkaBillingModel, "desired_kafka_billing_model", 1),
+			// ValidateKafkaUserFacingUpdateFields(ctx, h.authService, kafkaRequest, &kafkaUpdateReq), // TODO should some of this checks in here related to owner be performed?
```

Comment:
Should we keep some of the behavior of this method for this endpoint?

Reply:
Hmmm... Let's keep this part: https://github.com/bf2fc6cc711aee1a0c2a/kas-fleet-manager/blob/867982657339b85a34555e5bff3d5b34eaea7856/internal/kafka/internal/handlers/validation.go#L277-L289.

The only doubt I have is: shouldn't we do that in some auth middleware?

Reply:
Regarding doing that in some auth middleware. It is a possibilty.

I suspect one of the possible reasons to not being done in a middleware is due to some endpoints not needing it. That means that the endpoints in our API at code level should be restructured at subrouter level so that only applies to certain endpoints.

For now I will create a validation that performs that part. We can look at a general restructuring of that for KFM in general.

Reply:
I added a new commit with the changes.
The new commit:
1. Refactors the `ValidateKafkaUserFacingUpdateFields` by extracting the code that performs ownership check into its own new validator. The new validator is called `validateUserIsKafkaOwnerOrOrgAdmin`. `ValidateKafkaUserFacingUpdateFields` calls that other validator as part of its code
2. The kafka promote endpoint now calls `validateUserIsKafkaOwnerOrOrgAdmin` to ensure promotion is only allowed by the kafka owner or for an org admin of the kafka instance's organization